text/src/grep.rs

```diff
@@ -0,0 +1,253 @@
+        }
+
+        // Check if conflicting options are used together
+        if self.pattern_list.is_empty()
+            && self.pattern_file.is_empty()
+            && self.single_pattern_list.is_none()
+        {
+            return Err("Required at least one pattern list or file".to_string());
+        }
+
+        Ok(())
+    }
+
+    fn resolve_patterns(&mut self) -> Result<(), Box<dyn Error>> {
+        for pf in &self.pattern_file {
+            self.pattern_list.extend(Self::get_file_patterns(pf)?);
+        }
+
+        self.pattern_list = self
+            .pattern_list
+            .iter()
+            .flat_map(|pattern| pattern.split('\n').map(String::from))
+            .collect();
+
+        match &self.single_pattern_list {
+            // if single_pattern_list is none, then pattern_list is not empty
+            None => {}
+            // single_pattern_list might get files value
+            Some(pattern) => {
+                if !self.pattern_list.is_empty() {
+                    // pattern_list is not empty, then single_pattern_list took files value
+                    self.files.insert(0, PathBuf::from_str(pattern.as_str())?);
+                } else {
+                    // pattern_list is empty, then single_pattern_list is only pattern
+                    self.pattern_list = vec![pattern.to_string()]
+                }
+            }
+        }
+
+        self.regex_patterns = if !self.use_string {
+            self.pattern_list
+                .iter()
+                .map(|p| Regex::new(p).unwrap())
+                .collect()
+        } else {
+            vec![]
+        };
+
+        Ok(())
+    }
+
+    fn get_file_patterns<P: AsRef<Path>>(path: P) -> Result<Vec<String>, Box<dyn Error>> {
+        BufReader::new(File::open(&path)?)
+            .lines()
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|e| Box::new(e) as Box<dyn Error>)
+    }
+
+    fn match_patters(&self, input: &str) -> bool {
+        if self.use_string {
+            self.pattern_list.iter().any(|p| input.contains(p))
+        } else {
+            self.regex_patterns.iter().any(|r| r.is_match(input))
+        }
+    }
+}
+
+fn grep(args: &Args) -> Result<(), Box<dyn Error>> {
+    let mut matches_counter = 0;
+    if args.files.is_empty() {
+        let reader: Box<dyn BufRead> = Box::new(BufReader::new(io::stdin()));
+        process_input(args, "(standard input)", &mut matches_counter, reader)?;
+    } else {
+        for file in &args.files {
+            let reader: Box<dyn BufRead> = Box::new(BufReader::new(File::open(file)?));
+            process_input(
+                args,
+                file.display().to_string(),
+                &mut matches_counter,
+                reader,
+            )?;
+        }
+    }
+
+    if args.count {
+        println!("{matches_counter}");
+    }
+
+    Ok(())
+}
+
+fn process_input(
+    args: &Args,
+    source_name: impl Display,
+    matches_counter: &mut i32,
```

Comment:
matches_counter need to be Option. No need to count matches once again if args.count = false

Reply:
Refactored into `OutputMode` enum.